include/tvm/target/se_scope.h

```diff
@@ -0,0 +1,330 @@
+ *   => (kDLCPU, 3, "llvm", "local")
+ * \endcode
+ *
+ * These operations are needed during device planning.
+ *
+ */
+class SEScopeNode : public Object {
+ public:
+  /*!
+   * \brief The \p DLDeviceType of the device. If \p target is known then this will be equal to
+   * \p target->kind->device_type. If \p target is null then the target is to be determined by
+   * a later pass.
+   *
+   * This is needed to support the legacy "on_device" and "device_copy" calls which only allow
+   * a \p DLDeviceTypes (as an integer) to be given.
+   *
+   * kInvalidDeviceType denotes unconstrained.
+   */
+  DLDeviceType device_type() const { return device_type_; }
+
+  /*!
+   * \brief The 'virtual' device identifier for the device. This must be resolved to a physical
+   * device identifier either during compilation or at runtime.
+   *
+   * -1 denotes unconstrained. May be 0 if not significant.
+   */
+  int virtual_device_id() const { return virtual_device_id_; }
+
+  /*!
+   * \brief The \p Target describing how to compile for the device.
+   *
+   * Null denotes unconstrained (though if device_type is known then only a target of that
+   * type is allowed).
+   */
+  const Target& target() const { return target_; }
+
+  /*!
+   * \brief The scope of memory within the device.
+   *
+   * Empty denotes unconstrained.
+   */
+  const MemoryScope& memory_scope() const { return memory_scope_; }
+
+  /*!
+   * \brief Returns true if scope is fully unconstrained, ie no target/device type, virtual device
+   * id or memory scope is specified.
+   */
+  bool is_fully_unconstrained() const {
+    return !target_.defined() && device_type_ == kInvalidDeviceType && virtual_device_id_ == -1 &&
+           memory_scope_.empty();
+  }
+
+  /*!
+   * \brief Returns true if scope is fully constrained, ie target, virtual device id and
+   * memory scope are all specified.
+   */
+  bool is_fully_constrained() const {
+    return target_.defined() && virtual_device_id_ != -1 && !memory_scope_.empty();
+  }
+
+  Device ToDevice() const {
+    ICHECK(device_type_ != kInvalidDeviceType);
+    ICHECK(virtual_device_id_ != -1);
+    Device device;
+    device.device_type = device_type_;
+    device.device_id = virtual_device_id_;
+    return device;
+  }
+
+  void VisitAttrs(AttrVisitor* v);
+
+  // We implement structural equality (which reduces Targets to their str representation)
+  // so that unit tests can compare actual and expected SEScopes.
+  bool SEqualReduce(const SEScopeNode* other, SEqualReducer equal) const;
+  void SHashReduce(SHashReducer hash_reduce) const;
+
+  static constexpr const char* _type_key = "SEScope";
+  static constexpr const bool _type_has_method_sequal_reduce = true;
+  static constexpr const bool _type_has_method_shash_reduce = true;
+  TVM_DECLARE_FINAL_OBJECT_INFO(SEScopeNode, Object);
+
+ private:
+  // We keep the fields private so the constructor memoization can't be upset by mutation.
+  DLDeviceType device_type_;
```

Comment:
Thanks for that. I was worried about accidental mutation given the memoization but you're quite right safety is via the const ptrs rather than the field access.